File: LargestRectangle/LargestRectangle.cpp
// LargestRectangle.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <vector>

int GetArea(const std::vector<int>& heights, size_t index)
{
    int minHeight = heights[index];
    int area = heights[index];
    for (size_t i = index+1; i < heights.size(); ++i)
    {
        if (heights[i] == 0)
            break;

        minHeight = std::min(minHeight, heights[i]);
        area = std::max(area, minHeight * static_cast<int>(i - index + 1));
    }

    return area;
}

int GetLargestArea(std::vector<int>& heights)
{
    int area = 0;
    for (size_t i = 0; i < heights.size(); ++i)
    {
        area = std::max(area, GetArea(heights, i));
    }

    return area;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of heights: ";
        std::cin >> count;
        if (count == 0)
            break;

        std::vector<int> vec(count);
        std::cout << "The heights: ";
        for (int i = 0; i < count; ++i)
            std::cin >> vec[i];

        std::cout << "The largest area is " << GetLargestArea(vec) << std::endl;
    }
}

File: Go/LargestRectangle.go
package main

import "fmt"

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetArea(heights []int, index int) int {
	area := heights[index]
	height := heights[index]
	for i := index + 1; i < len(heights); i++ {
		height = min(height, heights[i])
		if height == 0 {
			break
		}
		area = max(area, (i-index+1)*height)
	}

	return area
}

func GetLargestArea(heights []int) int {
	maxArea := 0

	for i := 0; i < len(heights); i++ {
		maxArea = max(GetArea(heights, i), maxArea)
	}

	return maxArea
}

func main() {
	for {
		fmt.Print("Number of heights: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		heights := GetArrayFromInput(count, "Please enter the heights: ")
		area := GetLargestArea(heights)
		fmt.Print("The largest rectangle in the histogram is ")
		fmt.Println(area)
	}
}
